packages/postgraphile-core/__tests__/unit/read-cache.test.js

```diff
@@ -0,0 +1,84 @@
+const fs = require("fs");
+const graphileBuild = require("graphile-build");
+const { getPostGraphileBuilder } = require("../..");
+
+jest.mock("fs");
+jest.mock("graphile-build");
+
+beforeEach(() => {
+  // required for mocks assertions
+  jest.resetAllMocks();
+});
+
+/*
+
+Current behaviour of getPostGraphileBuilder and readCache option:
+
+1. cacheString is loaded by reading the file defined by readCache
+2. memoizeCache is set by JSON.parse(cacheString)
+3. persistentMemoizeWithKey is a function created dynamically that will use memoizeCache
+4. getBuilder(imported from graphile-build) is called, passing persistentMemoizeWithKey as one of the arguments
+
+Test strategy for readCache
+
+1. Mock fs.readFile to control input
+2. Mock getBuilder to control output
+3. Test different inputs and validate that getBuilder is called with expected output
+
+*/
+
+test("when cache file has empty object, persistentMemoizeWithKey should be a function", async () => {
+  // mock fs.readFile to return an object
+  fs.readFile.mockImplementationOnce((path, options, cb) => cb(null, "{}"));
+
+  // mock getBuilder to fake output
+  const expectedOuput = {};
+  graphileBuild.getBuilder.mockResolvedValueOnce(expectedOuput);
+  // call our method and test output
+  const output = await getPostGraphileBuilder({}, [], {
+    readCache: "path/to/cache",
+  });
+  expect(output).toBe(expectedOuput);
```

Comment:
Since this assertion is the same as the one on line 57, rather than an empty object could we define something like `{"__test": true}` and assert that matches? Thanks! :pray:

Reply:
Sure, I'll take a look tomorrow

Reply:
Hey @benjie I've reviewed your comment but I'm not sure what you mean. The check:
```javascript
const output = await getPostGraphileBuilder({}, [], {});
expect(output).toBe(expectedOuput);
```
its not about the value of `expectedOutput` but rather the fact that we got this output. This value is coming from [here](https://github.com/graphile/graphile-engine/blob/master/packages/postgraphile-core/src/index.ts#L365). Because this is a unit test, we are mocking `getBuilder` which is from another package:
```js
  // mock getBuilder to fake output
  const expectedOuput = {};
  graphileBuild.getBuilder.mockResolvedValueOnce(expectedOuput);
```
So which value we put there is completely irrelevant. Is only about checking that  `getPostGraphileBuilder` got to the end. Given than we test the important outcome (that we pass the `persistentMemoizedWithKey` to `getBuilder`) in the next lines, we could completely remove the  `expectedOutput` check
```js
expect(graphileBuild.getBuilder).toHaveBeenCalledTimes(1);
  // check persistentMemoizeWithKey, the actual "result" of readCache flag
  const {
    persistentMemoizeWithKey,
  } = graphileBuild.getBuilder.mock.calls[0][1];
  expect(persistentMemoizeWithKey).toBeUndefined();
```
Either we remove it or we keep it as is, but changing the mocked value will not provide us with better checks (we would be testing jest mock functions hehe)

Reply:
Sorry yeah, this was the wrong place to add this suggestion - leave this line as-is, and change line 32 to return `'{"__test": true}'` instead of `"{}"`.

We should then call the generated `persistentMemoizeWithKey` at the bottom of the test with a know value and assert that it matches the expected value based on the cache data we've fed in, e.g.:

```js
expect(persistentMemoizeWithKey('__test')).toEqual(true);
expect(() => persistentMemoizeWithKey('unknown_key')).toThrow();
```

Reply:
aahh makes sense, it is a good check. I'll add it